core/workflow-compiling-service/src/main/scala/edu/uci/ics/amber/compiler/model/LogicalPlan.scala

```diff
@@ -1,39 +1,35 @@
 package edu.uci.ics.amber.compiler.model
 
 import com.typesafe.scalalogging.LazyLogging
+import edu.uci.ics.amber.compiler.model.LogicalPlan.LogicalEdge
 import edu.uci.ics.amber.core.storage.FileResolver
-import edu.uci.ics.amber.operator.LogicalOp
-import edu.uci.ics.amber.operator.source.scan.ScanSourceOpDesc
 import edu.uci.ics.amber.core.virtualidentity.OperatorIdentity
 import edu.uci.ics.amber.core.workflow.PortIdentity
-import org.jgrapht.graph.DirectedAcyclicGraph
-import org.jgrapht.util.SupplierUtil
+import edu.uci.ics.amber.operator.LogicalOp
+import edu.uci.ics.amber.operator.source.scan.ScanSourceOpDesc
+import scalax.collection.OneOrMore
+import scalax.collection.generic.{AbstractDiEdge, MultiEdge}
+import scalax.collection.mutable.Graph
 
-import java.util
 import scala.collection.mutable.ArrayBuffer
 import scala.util.{Failure, Success, Try}
 
 object LogicalPlan {
 
-  private def toJgraphtDAG(
+  case class LogicalEdge(logicalLink: LogicalLink)
+      extends AbstractDiEdge(logicalLink.fromOpId, logicalLink.toOpId)
+      with MultiEdge {
+    override def extendKeyBy: OneOrMore[Any] =
+      OneOrMore.one((logicalLink.fromPortId, logicalLink.toPortId))
+  }
+
+  private def toScalaDAG(
       operatorList: List[LogicalOp],
       links: List[LogicalLink]
-  ): DirectedAcyclicGraph[OperatorIdentity, LogicalLink] = {
```

Comment:
Does the `Graph` class ensure the plan is a DAG? For jgrapht if a plan is not a DAG an a error will be thrown. If Graph does not ensure the acyclicity part of a DAG, do we implement that ourselves? The acyclicity is crucial for scheduling (`RegionDAG`), which is currently still implemented as a jgrapht DAG.